### DISCUSSION TITLE
ESP32 Hardware Serial

### DISCUSSION THREAD

**Discussion by trlafleur** [arendst/Tasmota#15461] [Q&A] [answered]
How do you tell Tasmota that you're using a hardware Serial port??
I'm using serial RX on pin GPIO16 and TX on GPIO17 on an Adafruit ESP32 Feather.
This should be hardware serial port 2...

But in my program, I do an " if (D6100Serial->hardwareSerial())" , but it never executes...

My program runs fine, but I'm curious about the proper use and selection of hardware serial ports...

~~~
/* ******************************************************** */
void D6100Init(void) 
{
  if (PinUsed(GPIO_D6100_RX) && PinUsed(GPIO_D6100_TX)) 
  {
    D6100Serial = new TasmotaSerial(Pin(GPIO_D6100_RX), Pin(GPIO_D6100_TX), 2, 0, D6100_BUFFER_SIZE);
    if (D6100Serial) 
    {
      if (D6100Serial->begin(D6100_BAUDRATE)) 
      {
        if (D6100Serial->hardwareSerial()) 
        { 
          ClaimSerial();                            // Tell Tasmota we have serial port
          AddLog(LOG_LEVEL_INFO, PSTR("%s"), "*** D6100 Claiming Hardware Serial Port");
        }       
      }
    }
  }
}   // end of: void D6100Init(void)
~~~

~~~
Showing logs:
[13:50:58]
[13:50:58]00:00:00.003-212/50 HDW: ESP32-D0WD-V3 
[13:50:58]./components/esp_littlefs/src/littlefs/lfs.c:1071:error: Corrupted dir pair at {0x0, 0x1}
[13:50:59]00:00:00.459-210/49 UFS: FlashFS mounted with 312 kB free
[13:50:59]00:00:00.466-210/49 CFG: Use defaults
[13:50:59]00:00:00.634 QPC: Reset
[13:50:59]00:00:00.743 BRY: Berry initialized, RAM used=3800 bytes
[13:50:59]00:00:00.775 Project  - Tasmota Version 11.0.0.3(tasmota)-2_0_2_3(2022-04-25T13:49:29)
[13:50:59]00:00:00.825 RSL: RESULT = {"SetOption8":"ON"}
[13:51:00]00:00:01.006 WIF: Connecting to AP1 lafleur in mode 11n as 6C4EFC-3836...
[13:51:05]00:00:06.052 WIF: Connected
[13:51:06]00:00:06.879 HTP: Web server active on 6C4EFC-3836 with IP address 192.94.166.53
[13:51:09]12:51:09.555 MQT: Attempting connection...
[13:51:10]12:51:10.326 MQT: Connected
[13:51:10]12:51:10.334 MQT: RSF/TASMOTA/6C4EFC/LWT = Online
[13:51:10]12:51:10.339 MQT: RSF/TASMOTA/6C4EFC/cmnd/POWER = 
[13:51:10]12:51:10.361 MQT: RSF/TASMOTA/6C4EFC/INFO1 = {"Info1":{"Module":"D6100 Alarm","Version":"11.0.0.3(tasmota)","FallbackTopic":"cmnd/DVES_6C4EFC_fb/","GroupTopic":"RSF/TASMOTA/tasmotas/cmnd/"}}
[13:51:10]12:51:10.379 MQT: RSF/TASMOTA/6C4EFC/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"6C4EFC-3836","IPAddress":"192.94.166.53"}}
[13:51:10]12:51:10.399 MQT: RSF/TASMOTA/6C4EFC/INFO3 = {"Info3":{"RestartReason":"Vbat power on reset","BootCount":1}}
[13:51:14]12:51:14.383 MQT: RSF/TASMOTA/6C4EFC/STATE = {"Time":"2022-04-25T12:51:14","Uptime":"0T00:00:13","UptimeSec":13,"Heap":155,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":84,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":42},"Wifi":{"AP":1,"SSId":"lafleur","BSSId":"9C:C9:EB:9A:56:C1","Channel":6,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:06"}}
[13:51:14]12:51:14.439 MQT: RSF/TASMOTA/6C4EFC/SENSOR = {"Time":"2022-04-25T12:51:14","ESP32":{"Temperature":126.0},"TempUnit":"F"}

~~~

**Comment by barbudor** [comment, marked as answer]
On ESP32, it's always hw serial
See this driver https://github.com/arendst/Tasmota/blob/f66574ca55d4c20c92331d8c3b7ae3bed87ba9a0/tasmota/xnrg_15_teleinfo.ino#L645

**Comment by barbudor** [comment]
And hardwareSerial() returns true only for serial 0, the one used for console
https://github.com/arendst/Tasmota/blob/f66574ca55d4c20c92331d8c3b7ae3bed87ba9a0/lib/default/TasmotaSerial-3.5.0/src/TasmotaSerial.cpp#L224

**Comment by trlafleur** [comment]
why is that?? just curious??



~~ _/) ~~~~ _/) ~~~~ _/) ~~~~ _/) ~~

Tom Lafleur

**Reply by barbudor** [reply]
The behavior is inherited of esp8266
On esp8266 the only one hw serial is used for the console by default. So when a driver uses it, it needs to claim it.

On ESP32, the console uses uart0 and TasmotaSerial start to allocate from uart2 (on ESP32, or uart1 on C3). Beside TasmotaSerial do not support software Serial on ESP32
So there is no need to claim the serial
Which means your code works fine and as expected as it is.

In fact, it is the other driver that may have a problem if it gets allocates uart0.

**Comment by hvegh** [comment]
> On ESP32, the console uses uart0 and TasmotaSerial start to allocate from uart2 (on ESP32, or uart1 on C3).

Does that mean only UART1 is usable in Tasmota on an esp32c3?
The [SerialBridge](https://tasmota.github.io/docs/Commands/#serial-bridge) commands seam not able to specify a specific UART??


I have an esp32c3 in cdc mode, meaning console logging is done via the internal USB port of the ESP. 
The device has 2 hardware uarts: UART0 on GPIO0,1 and UART1 on GPIO8,9.

Is there a way I can use both ports and configure them separately?

**Reply by arendst** [reply]
Try it and see what happens.

BTW there is no relation between UART and GPIO. Any GPIO can use hardware serial.

**Reply by hvegh** [reply]
I will check this!

In order to interface to the P1 port of the energy meter directly without using an inverter, I am currently trying to add a
boolean command option **SerialInvert**  which should be propagated to the HardwareSerial layer.
 
I tested this with HardwareSerial code with invert=true  and works great, just using a standard  RJ14 plug and an external pull-up resistor. 
(You could probably even lose the pull-up resistor using the internal pull-up of the GPIO port)

Is this something that could be part of the mainline? 
I would be happy to submit a pull-request for comments...

**Reply by arendst** [reply]
Try latest dev branch. I enabled invert for HardwareSerial on both ESP8266 and ESP32.

Use a call like 
```
P1Serial = new TasmotaSerial(Pin(GPIO_P1_RX), Pin(GPIO_P1_TX), 2, 0, MIN_INPUT_BUFFER_SIZE, 1); 
```

**Comment by hvegh** [comment]
Disclaimer, I have just started to explore the possibilities of the Tasmota platform, so I don't always know what I am doing. 

I have tested two devices:

[lolin32 with oled](https://i0.wp.com/randomnerdtutorials.com/wp-content/uploads/2019/07/lolin32-oled-pinout.jpg)
My simple code uses:
```
#define PIN_UART_RX 16 // UART2
#define PIN_UART_TX 17 // UART2
SerialP1.begin(115200, SERIAL_8N1, PIN_UART_RX, PIN_UART_TX, true);
```
Then it happily receives the OBIS ASCII telegram with the _transmit enable_ port is hard wired to the 5 Volt from the USB.

On Tasmota I have assigned the RX_ser and RX_bridge to GPIO 16 and hardcoded the invert flag in `xdrv_08_serial_bridge.ino`:
`+    SerialBridgeSerial = new TasmotaSerial(Pin(GPIO_SBR_RX), Pin(GPIO_SBR_TX), HARDWARE_FALLBACK, 0, MIN_INPUT_BUFFER_SIZE, true);  // 256
`
Now looking back at the driver code I might have failed to set TX pin to GPIO 17... so the TasmotaSerial does not get instantiated, still might try that..  

Next thing I tried, was to modify user_config_override.h: 
```
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
/*
 * tasmota_xsns_sensor/xsns_53_sml.ino - SML,OBIS,EBUS,RAW,COUNTER interface for Tasmota
 *
 * 3601:mp->meter_ss->begin(mp->params, smode, mp->srcpin, mp->trxpin, mp->so_flags.SO_TRX_INVERT);

typedef union {
  uint8_t data;
  struct {
    uint8_t SO_DWS74_BUG : 1;
    uint8_t SO_OBIS_LINE : 1;
    uint8_t SO_TRX_INVERT : 1;
  };
} SO_FLAGS;

> bit 0 -> SO_TRX_INVERT

https://tasmota.github.io/docs/Smart-Meter-Interface/#special-commands

>D
>B
->sensor53 r
>M 1
+1,16,o,1,115200,P1
1,0-1:24.2.3(@#),Timestamp Gas,,timestamp_gas,0
1,1-0:1.8.1(@1,Tariff High,kWh,Tarif1,3
1,1-0:1.8.2(@1,Tariff Low,kWh,Tarif2,3
1,0-0:96.14.0(@1,Tariff,,tariff,0
1-0:1.7.0(@1,Power,kW,Power,3
1,1-0:32.7.0(@1,Voltage L1,V,Voltage,1
1,1-0:52.7.0(@1,Voltage L2,V,Voltage,1
1,1-0:72.7.0(@1,Voltage L3,V,Voltage,1
1,1-0:31.7.0(@1,Current L1,A,Current,1
1,1-0:51.7.0(@1,Current L2,A,Current,1
1,1-0:71.7.0(@1,Current L3,A,Current,1
#

 */
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif

```
Seems that the `xsns_53_sml.ino` code already had some provisions for inverted serial lines.
If I understand the code correctly the **1** in +1,16,o,**1**,115200,P1 ensures bit 0 of so_flags is set so the inversion is triggered.

Any ideas?

**Reply by arendst** [reply]
The SML code is maintained seperate from the Tasmota main code. You'll likely find features there that you cannot use in other parts of Tasmota.

That said, SML supports inverted signals in it's own way. It circumvents Tasmota by directly calling hardwareSerial() likely causing uart issues in other parts of Tasmota. 

If you think SML should change call @gemu2015. Otherwise using TasmotaSerial call me.

**Reply by gemu2015** [reply]
@hvegh 
according to SML options, read the docs about "special options" in the descriptor definition
